src/runtime/hexagon_dma_device_interface.cpp

```diff
@@ -0,0 +1,251 @@
+}}}}
+
+using namespace Halide::Runtime::Internal::Dma;
+
+//The get function is necessary because in some places like get memory
+//we do not have access to the inframe buffer
+void halide_hexagon_get_dma_context (void* user_context, p_dma_context* context) {
+    halide_assert(user_context, context != NULL);
+    *context = pdma_context;
+}
+
+void halide_hexagon_acquire_dma_context(void* user_context, p_dma_context* ctx, int num_of_frames, bool create=true) {
+    halide_assert(user_context, ctx != NULL);
+    // If the context has not been initialized, initialize it now.
+    if (!pdma_context && create) {
+        halide_hexagon_dmart_create_context(user_context, &pdma_context, num_of_frames);
+    }
+    *ctx = pdma_context;
+}
+
+namespace Halide { namespace Runtime { namespace Internal { namespace Dma {
+
+HexagonDmaContext::HexagonDmaContext(void* user_context): user_ctxt(user_context),
+                           context(NULL) {
+
+    halide_hexagon_acquire_dma_context(user_ctxt, &context, 1);
+    halide_assert(user_context, context != NULL);
+}
+
+HexagonDmaContext::HexagonDmaContext(void* user_context, int num_of_frames): user_ctxt(user_context),
+                          context(NULL) {
+
+    halide_hexagon_acquire_dma_context(user_ctxt, &context, num_of_frames);
+    halide_assert(user_context, context != NULL);
+}
+
+HexagonDmaContext::~HexagonDmaContext() {
+    user_ctxt = 0;
+    context = 0;
+}
+}}}}
+
+using namespace Halide::Runtime::Internal::Dma;
+
+extern "C" {
+
+extern WEAK halide_device_interface_t hexagon_dma_device_interface;
+
+WEAK int halide_hexagon_dma_device_release(void *user_context) {
+    return HEX_SUCCESS;
+}
+
+WEAK int halide_hexagon_dma_device_malloc(void *user_context, halide_buffer_t *buf) {
+
+    //////////////////////////////////////////////////////////////////////
+    p_dma_context dma_handle = reinterpret_cast<p_dma_context>(buf->device);
+    uintptr_t  frame = halide_hexagon_dmart_get_frame(user_context, dma_handle);
+    //////////////////////////////////////////////////////////////////////////
+
+    //Check for dma_allocation
+    bool dma_allocate;
+    void* handle = 0;
+    if (halide_hexagon_dmart_allocate_dma(user_context, pdma_context, frame, &dma_allocate) == -1) {
+        error(user_context) << "Undefined Error";
+        return HEX_ERROR;
+    }
+
+    if (dma_allocate) {
+        handle = dma_allocate_dma_engine();
+        if (handle == 0) {
+            error(user_context) << "halide_hexagon_dma_device_malloc:Failed to allocate the read DMA engine.\n";
+            return HEX_ERROR;
+        }
+
+        if (halide_hexagon_dmart_set_dma_handle(user_context, pdma_context, handle, frame)) {
+            error(user_context) << "halide_hexagon_dma_device_malloc:Function failed to set DMA Handle to DMA context \n";
+            return HEX_ERROR;
+        }
+    }
+    return 0;
+}
+
+WEAK int halide_hexagon_dma_device_free(void *user_context, halide_buffer_t *buf) {
```

Comment:
This function makes no reference to 'buf' other than to assert it's not NULL. This function should be freeing the device allocation of buf.